### ISSUE TITLE
feature: Create a tagged input control

### ISSUE THREAD

**Issue by Brocco** [ng-bootstrap/ng-bootstrap#532]
A tagged input combines the features of a type-ahead with multiple selections.

An implementation (style, not technical details) like what is found here would be ideal https://bootstrap-tagsinput.github.io/bootstrap-tagsinput/examples/ 

![image](https://cloud.githubusercontent.com/assets/1565117/17344232/63e44c60-58cf-11e6-9289-505f7e20191c.png)

The current discussion of this issue should surround whether control like this should be included in this repo/project. Implementation/technical details would follow.

**Comment by icfantv** [comment]
@Brocco this is already under discussion for the typeahead component similar to what Material has done.  As mentioned in the original typeahead component discussion, this would also need a virtual scroller directive to prevent the entire dropdown from being re-rendered every time it is opened (as is currently the issue with ui-select and is a huge performance problem).

This said, we need to get modal and datepicker working first before working on an extended feature set.  If you would like to start on a design document, you can look at [the one we used for the time picker](https://github.com/ng-bootstrap/ng-bootstrap/issues/24) as a launching point.

**Comment by pkozlowski-opensource** [comment]
@Brocco @icfantv so I was toying with the idea today and it seems like it would be _extremely_ easy to build a custom tagged input on top of the existing typeahead: https://plnkr.co/edit/sZNw1lO2y3ZZR0GxLyjD?p=preview

In the above plunker the tagged-input specific code amounts to ~7 LOC. CSS would need more toying with, I didn't spend too much time on it. 

Given the above I wonder if this makes sense to encapsulate those ~7 LOC in a specific directive. I kind of like the raw approach from the plunker as it gives people so much flexibility. Thoughts?

In any case, dumping it here so others can take the idea further.

**Comment by icfantv** [comment]
@pkozlowski-opensource this is awesome.  i had no idea it would be so straightforward.  RE making a separate directive, that might make the most amount of sense.  I think if we combined them, we'd also have to specify a single vs. multiple-select flag and IMO would be nice to not pollute the typeahead component with that.

**Comment by Brocco** [comment]
I am using the typeahead for something similar but due to layout constraints I didn't have room for tags, so instead I populate them into an attached dropdown. Add an item, it goes into the drop down, click on an item in the dropdown to remove it, with a "Remove All" option at the bottom of the list. Might be a nice idea to combine both UI patterns into a single control?

FYI, in mine instead of getting a reference to the input via ViewChild, I pass it back to the component as an argument to the `selectItem` event callback
`#myinput (selectItem)="selectHandler($event, myInput)"`

**Comment by ravtakhar** [comment]
Is this still a feature that is being considered?

**Comment by pkozlowski-opensource** [comment]
@ravtakhar have you seen https://github.com/ng-bootstrap/ng-bootstrap/issues/532#issuecomment-248750676 ?

What else would you expect?

**Comment by ravtakhar** [comment]
I looked through the example and it's very lean. It's very similar to something that I ended up building for labels.

I saw the issue open and thought that someone might be working on a more complete/cleaner version. The biggest thing I want to see is when you click the box (in this case typeahead) a dropdown to be presented with a list of options to pick from. Now that I think about it... this is I guess should be an extension to typeahead and not a completely new component.

**Comment by pkozlowski-opensource** [comment]
> I saw the issue open and thought that someone might be working on a more complete/cleaner version.

No, we are not actively working on another version as so far we didn't comp up with APIs that would improve ergonomics of the solution from the plunker _and_ keep reasonable flexibility. But I'm sure that there are ways of making it all easier so if you've got proposal(s), please speak up.

 > Now that I think about it... this is I guess should be an extension to typeahead and not a completely new component.

Yes. I would say a new directive / component that _uses_ typeahead. But yeh, hard to say without more concrete proposal(s).

**Comment by intellix** [comment]
Noticed the plnkr stopped working so created an updated Stackblitz: https://stackblitz.com/edit/angular-rrjczw
